Implement custom url helpers and polymorphic mapping

pixeltrix opened the pull request:
Implement the features outlined in #22512.
- [x] Add support for defining custom url helpers in routes.rb
- [x] Add support for  defining custom polymorphic mappings in routes.rb

pixeltrix committed:
Wrap routes.url_helpers.url_for via a proxy

The singleton url_for on Rails.application.routes.url_helpers isn't the
same as the url_for you get when you include the module in your class as
the latter has support for polymorphic style routes, etc. whereas the
former accepts only a hash and is the underlying implementation defined
on ActionDispatch::Routing::RouteSet.

This commit changes the singleton method to call through a proxy instance
so that it gets the full range of features specified in the documentation
for url_for.

pixeltrix committed:
Add support for defining custom url helpers in routes.rb

Allow the definition of custom url helpers that will be available
automatically wherever standard url helpers are available. The
current solution is to create helper methods in ApplicationHelper
or some other helper module and this isn't a great solution since
the url helper module can be called directly or included in another
class which doesn't include the normal helper modules.

Reference #22512.

pixeltrix commented:
@rails/core pushing up a WIP for some feedback on method name and implementation.

rafaelfranca commented on actionpack/lib/action_dispatch/routing/route_set.rb:
```diff
@@ -144,6 +159,62 @@ def length
+          def call(t, args, options, outer_options = {})
+            url_options = eval_block(t, args, options)
+
+            case url_options
```
Is not this duplicated somewhere else?

pixeltrix replied:
It's similar to the behaviour in ActionDispatch::Routing::UrlFor but not exactly the same, mainly because of the options being needed to be combined into the arguments for `url_for`.

dhh commented on actionpack/lib/action_dispatch/routing/mapper.rb:
```diff
@@ -2007,6 +2007,46 @@ def concerns(*args)
+        #
+        # NOTE: It is the url helper's responsibility to return the correct
+        # set of options to be passed to the `url_for` call.
+        def url_helper(name, options = {}, &block)
```
Not liking the word `url_helper` for this. It doesn't fit well with the rest of the metaphor we're using with drawing etc. What didn't you like about `direct` from the original pitch?

```
direct :homepage do
  "http://www.rubyonrails.org"
end

direct :commentable do |model|
  [ model, anchor: model.dom_id ]
end

direct :main do
  { controller: 'pages', action: 'index', subdomain: 'www' }
end
```

Looks and sounds great to me.

dhh replied:
@pixeltrix Did you have any reservations about using `direct`? Would love to get this merged before rc1. Also, we can use `lookup` for the `link_to @model` change.

pixeltrix committed:
Rename url_helper to direct

pixeltrix committed:
Only accept symbols and strings for Mapper#direct

pixeltrix committed:
Don't allocate a hash unnecessarily

pixeltrix committed:
Add test for calling a url helper in Mapper#direct

pixeltrix committed:
Add custom polymorphic mapping

Allow the use of `direct` to specify custom mappings for polymorphic_url, e.g:

  resource :basket
  direct(class: "Basket") { [:basket] }

This will then generate the following:

  >> link_to "Basket", @basket
  => <a href="/basket">Basket</a>

More importantly it will generate the correct url when used with `form_for`.

Fixes #1769.

pixeltrix committed:
Add CHANGELOG entry for `direct` method

matthewd commented on actionpack/lib/action_dispatch/routing/mapper.rb:
Ensure `options` is nil, as we're ignoring it?

matthewd commented on actionpack/lib/action_dispatch/routing/route_set.rb:
```diff
@@ -143,6 +158,25 @@ def length
+
+          @path_helpers_module.module_eval do
+            define_method(:"#{name}_path") do |*args|
+              options = args.extract_options!
```
Is there any benefit to doing this extraction here? Seems like that could be `call`'s problem, where it already does so for nested arrays.

matthewd commented on actionpack/lib/action_dispatch/routing/route_set.rb:
```diff
@@ -537,6 +598,61 @@ def add_route(mapping, path_ast, name, anchor)
+            if url_options.permitted?
+              t.url_for(url_options.to_h.merge(outer_options))
+            else
+              raise ArgumentError, "Generating an URL from non sanitized request parameters is insecure!"
```
*a* URL

matthewd commented on actionpack/lib/action_dispatch/routing/route_set.rb:
```diff
@@ -537,6 +598,61 @@ def add_route(mapping, path_ast, name, anchor)
+        def call(t, args, options, outer_options = {})
+          url_options = eval_block(t, args, options)
+
+          case url_options
```
Can we push this whole `case` back into `t` (that's the "real" route thingy, right?), by just adding a new `url_for`-like method that takes an explicit `path_only` parameter?

pixeltrix replied:
Isn't there a `path_for` we could use?

matthewd commented on actionpack/test/dispatch/routing/direct_url_helpers_test.rb:
```diff
@@ -0,0 +1,215 @@
+
+    direct(class: "Basket") { |basket| [:basket] }
+    direct(class: "User", anchor: "details") { |user, options| [:profile, options] }
+    direct(class: "Video") { |video| [:media, { id: video.id }] }
```
Given a `Model#to_param`, this would work as `[:media, video]`, right?

pixeltrix replied:
No, that would map to `media_video_path(@video)` which doesn't exist

matthewd replied:
Ah... ouch.

So there's no way to really delegate to another route helper directly.

Could we make `media_path(video)` work, *and* have that magically behave like media_url or media_path depending on whether we're doing video_url or video_path, a bit like ActionMailer already does by forcing a full URL even when you ask for just the path?

pixeltrix replied:
> Could we make media_path(video) work, and have that magically behave like media_url or media_path depending on whether we're doing video_url or video_path, a bit like ActionMailer already does by forcing a full URL even when you ask for just the path?

Possibly, `url_for` takes a strategy argument but we'd have to intercept the call somehow

matthewd replied:
The easy option would seem to be to have two different objects we can do the `instance_exec` on.. I don't remember how AM achieves its version, though.

matthewd commented on actionpack/CHANGELOG.md:
````diff
@@ -1,3 +1,42 @@
+        ```
+
+        This generates the correct singular URL for the form instead of the default
+        resources member url, e.g. `/basket` vs. `/basket/:id`.
````
IMO we should come up with a different example here -- something that uses a parent route, perhaps -- both because it's desirable to show an argument-using block, and because I think `resource` should be registering the singular URL automatically (with some conflict resolution to avoid an incompatible change if you define both singular and plural resources for the same class).

pixeltrix replied:
Sorry, not entirely sure what you mean by this - can you illustrate with an example? Do you mean that calling `resource :basket` should automatically do `direct(class: 'Basket') { [:basket] }` behind the scenes?

matthewd replied:
> Do you mean that calling resource :basket should automatically do direct(class: 'Basket') { [:basket] } behind the scenes?

Yes.

pixeltrix replied:
I don't think this would be a good idea - if we add it via `resource` then we'd need to add an option to configure/disable it and the thought of touching that code again makes me 😱

pixeltrix commented on actionpack/lib/action_dispatch/routing/mapper.rb:
Probably - I'm sure we can detect that

pixeltrix commented on actionpack/lib/action_dispatch/routing/mapper.rb:
And what if it isn't - raise or just merge?

pixeltrix commented on actionpack/lib/action_dispatch/routing/polymorphic_routes.rb:
```diff
@@ -156,6 +164,11 @@ def polymorphic_path_for_action(action, record_or_hash, options)
 
+        def polymorphic_mapping(record)
+          return false unless record.respond_to?(:to_model)
+          _routes.polymorphic_mappings[record.to_model.model_name.name]
```
👍

pixeltrix commented on actionpack/lib/action_dispatch/routing/route_set.rb:
```diff
@@ -95,9 +96,23 @@ def clear!
+            url_name = :"#{helper}_url"
+
+            if @path_helpers_module.method_defined?(path_name)
+              @path_helpers_module.send :undef_method, path_name
```
Unsure - need to check why we used `undef_method` in the first place

pixeltrix replied:
It was changed in 0088b08dcaf16176c8f9364d1d786f0c3728d369 - @tenderlove any reason `undef_method` was used instead of `removed_method` ?

pixeltrix replied:
I'm going to assume that `undef_method` was used because `remove_possible_method` uses that internally for the reason outlined in 0244c0d8f339385a9b420a0565b17d327bf25b13

pixeltrix replied:
Given the fact that we should allow people to access inherited methods if they include url helpers and then change a conflicting route name then probably best to switch to `remove_method`

matthewd commented on actionpack/lib/action_dispatch/routing/mapper.rb:
I'd say raise ArgumentError

pixeltrix committed:
Raise an error if `direct` is inside a scope block

pixeltrix commented on actionpack/lib/action_dispatch/routing/mapper.rb:
Maybe we should always use a hash?, e.g:

``` ruby
direct(method: :rubyonrails) { "http://rubyonrails.org" }
direct(class: "Basket") { "/basket" }
```

Helps to clear up some naming confusion - @dhh wdyt?

pixeltrix committed:
Support mapping of non-model classes

dhh commented on actionpack/lib/action_dispatch/routing/mapper.rb:
Not a fan of that. Then I'd rather we use two separate method names. We could use `direct` for the URL declaration and `resolve` for polymorphics. So it's:

```ruby
direct(:rubyonrails) { "http://rubyonrails.org" }
resolve("Basket") { "/basket" }
```

pixeltrix committed:
Prefer remove_method over undef_method

Using `undef_method` means that when a route is removed any other
implementations of that method in the ancestor chain are inaccessible
so instead use `remove_method` which restores access to the ancestor.

pixeltrix committed:
Removed `model_name` method to prevent warning

pixeltrix committed:
Fix typo in exception message

pixeltrix committed:
Push option extract into call method

pixeltrix committed:
Split direct method into two

Use a separate method called `resolve` for the custom polymorphic
mapping to clarify the API.

pixeltrix committed:
Fix schema leakage from dirty_test.rb

The column information for the testings table was being cached
so clear the cache in the test teardown.

jrochkind commented on actionpack/CHANGELOG.md:
````diff
@@ -1,3 +1,48 @@
+    ``` ruby
+    resource :basket
+    direct(class: "Basket") { [:basket] }
+    ```
````
Is this a typo, it says "Add the `resolve`" method, but then the example uses `direct` not `resolve`.  Not sure if the example is valid as it is and is demonstrating `direct`, or if it's a typo and the example is demonstrating `resolve`.  Trying to figure out these new features.

georgeclaghorn replied:
It’s corrected in master.

pixeltrix replied:
It was almost fixed in master - it is now 😄

sunnyrjuneja commented:
@pixeltrix Hey, great work on this PR. Was just reviewing the RC to see what's new in Rails 5.1. Personally, I found your `direct` example to be very easy to understand and illustrative. However, I wasn't able to totally grok why and when for `resolve`. Just some feedback. Thanks again for your hard work.

adammiribyan commented:
@sunnyrjuneja the [tests contain](https://github.com/rails/rails/blob/de24b8cb9c38a02b521d762d4b8eef11f6a78978/actionpack/test/dispatch/routing/custom_url_helpers_test.rb#L99-L103) a quite illustrative example for `resolve`.
